### PR TITLE
Weekly allocation and asset support

### PR DESCRIPTION
**Pull request by saweber**
## What does this PR change?
* support for window and step to accept 'w' for 'week'
* prevent accumulateByWeek from producing an error is the data is already accumulated into a week

## Does this PR relate to any other PRs?
* yes; 
https://github.com/opencost/opencost/pull/1794
https://github.com/kubecost/kubecost-cost-model/pull/1315/files

## How will this PR impact users?
* 

## Does this PR address any GitHub or Zendesk issues?
* Closes ...

## How was this PR tested?
* 

## Does this PR require changes to documentation?
* 

## Have you labeled this PR and its corresponding Issue as "next release" if it should be part of the next Opencost release? If not, why not?
*

### CODE REVIEW

File: pkg/kubecost/allocation.go [hunk 1]
```diff
@@ -2,6 +2,7 @@ package kubecost
 
 import (
 	"fmt"
+	"github.com/opencost/opencost/pkg/util/timeutil"
```
Comment (ameijer) [inline pkg/kubecost/allocation.go]: consider grouping this with the other opencost imports right below?

File: pkg/util/timeutil/timeutil.go [hunk 2]
```diff
@@ -267,13 +268,19 @@ func FormatDurationStringDaysToHours(param string) (string, error) {
...
+	}
 	daysFromMonday := int(date.Weekday())
 	return date.Add(-1 * time.Duration(daysFromMonday) * Day)
 }
 
 // RoundToStartOfFollowingWeek creates a new time.Time for the following Monday 00:00 UTC
 func RoundToStartOfFollowingWeek(t time.Time) time.Time {
 	date := time.Date(t.Year(), t.Month(), t.Day(), 0, 0, 0, 0, time.UTC)
+	if date.Weekday() == time.Sunday {
+		return date
```
Comment (ameijer) [inline pkg/util/timeutil/timeutil.go]: @saweber if I am trying to use this func on a sunday afternoon date, do we want it to return sunday 7 days from now?

File: pkg/util/timeutil/timeutil.go [hunk 3]
```diff
@@ -267,13 +268,19 @@ func FormatDurationStringDaysToHours(param string) (string, error) {
...
 func RoundToStartOfWeek(t time.Time) time.Time {
 	date := time.Date(t.Year(), t.Month(), t.Day(), 0, 0, 0, 0, time.UTC)
+	if date.Weekday() == time.Sunday {
+		return date
+	}
 	daysFromMonday := int(date.Weekday())
 	return date.Add(-1 * time.Duration(daysFromMonday) * Day)
 }
 
 // RoundToStartOfFollowingWeek creates a new time.Time for the following Monday 00:00 UTC
```
Comment (ameijer) [inline pkg/util/timeutil/timeutil.go]: should we update these comments/var names? Weekday() has the following underlying implementation: 
```
const (
	Sunday Weekday = iota
	Monday
	Tuesday
	Wednesday
	Thursday
	Friday
	Saturday
)
```
... I think this is returning sunday dates, not mondays?

**Review by ameijer — COMMENTED** [review]
hi @saweber took a quick pass - what do you think?

File: pkg/kubecost/allocation.go (hunk 1 above)
Comment (saweber) [reply pkg/kubecost/allocation.go]: Good catch! There was no reason for it to be there - other than is where IntelliJ autoimported it. :)

File: pkg/util/timeutil/timeutil.go (hunk 3 above)
Comment (saweber) [reply pkg/util/timeutil/timeutil.go]: Another good catch! Bad naming on my part, though I believe the functionality is correct (will validate + add tests).

File: pkg/util/timeutil/timeutil.go (hunk 2 above)
Comment (saweber) [reply pkg/util/timeutil/timeutil.go]: Good question. The reason for this is to prevent a window=1w/step=1w query for a week into the future.

Some additional unit tests to ensure correct edge case functionality. The intended functionality is that '1w' means the current week, with additional weeks calculating backwards from there.

File: pkg/kubecost/allocation.go (hunk 1 above)
Comment (nikovacevic) [reply pkg/kubecost/allocation.go]: We gotta get the IntelliJ crew on `gofmt`

Comment (nikovacevic) [reply pkg/kubecost/allocation.go]: I have no idea why that IDE uses a different formatter, but `gofmt` is the standard, so the two camps war over imports.

File: pkg/kubecost/window.go [hunk 4]
```diff
@@ -22,8 +22,8 @@ const (
 )
 
 var (
-	durationRegex       = regexp.MustCompile(`^(\d+)(m|h|d)$`)
-	durationOffsetRegex = regexp.MustCompile(`^(\d+)(m|h|d) offset (\d+)(m|h|d)$`)
+	durationRegex       = regexp.MustCompile(`^(\d+)(m|h|d|w)$`)
+	durationOffsetRegex = regexp.MustCompile(`^(\d+)(m|h|d|w) offset (\d+)(m|h|d)$`)
```
Comment (nikovacevic) [inline pkg/kubecost/window.go:26]: If we're accepting `w` for durations, we should probably accept it for offsets, too.

Comment (nikovacevic) [inline pkg/kubecost/window.go]: If we're accepting w for durations, we should probably accept it for offsets, too.

File: pkg/util/timeutil/timeutil.go (hunk 2 above)
Comment (nikovacevic) [reply pkg/util/timeutil/timeutil.go]: Will echo Alex's concern here -- I would expect this function to return "the following Sunday" if I gave it a Sunday at noon. We should remember that this is a very general package, so others may use these functions for other reasons than the ones we have in mind. Will suggest we change this back, unless there's more context I'm missing.

**Review by nikovacevic — CHANGES_REQUESTED** [review]

File: pkg/kubecost/allocation.go (hunk 1 above)
Comment (saweber) [reply pkg/kubecost/allocation.go]: IntelliJ gets thrown off by the whitespace line - and wants to alphabetize. I’ll make a note and figure out how to configure IntelliJ (and get it in here and KCM)

File: pkg/kubecost/window.go (hunk 4 above)
Comment (saweber) [reply pkg/kubecost/window.go]: @Sean-Holcomb has some specific concerns related to allowing an offset measured in weeks.

**Comment by ameijer** [comment]
@saweber I think the goimports tool can handle the grouping/ordering for you. Perhaps you can configure it to work with intellij?

-------- Original Message --------

File: pkg/kubecost/allocation.go (hunk 1 above)
Comment (nikovacevic) [reply pkg/kubecost/allocation.go]: I think Intellij has a `gofmt` option. To format code on save. That should do it.

Comment (saweber) [reply pkg/kubecost/allocation.go]: IntelliJ was already set to `go fmt`. For some reason, JetBrains' defaults are different than VSCode.

These settings seem to match y'alls preferences. I do not care how imports are sorted - let me know if this is acceptable and if we want to open separate PRs for OC/KCM to store this as project files so we don't have to repeat this for every IntelliJ/GoLand user.
Screenshot 2023-03-28 at 08 51 51

Comment (ameijer) [reply pkg/kubecost/allocation.go]: that looks great! and yes I think committing these to a file is a great idea.

File: pkg/kubecost/window.go (hunk 4 above)
Comment (Sean-Holcomb) [reply pkg/kubecost/window.go]: Not so much a concern as it seems like something we want to enable in the prometheus context, my point on line 26 was that `w` should not be an option on for both because this regex is only used for prometheus formatting. At this point I think accepting it for offset is fine for the sake of consistency.

Comment (saweber) [reply pkg/kubecost/window.go]: Discussed with Sean, we will go ahead and add it.

**Review by nikovacevic — APPROVED** [review]
